### ISSUE TITLE
Kanboard SAML SSO

### ISSUE THREAD

**Issue by trajche** [kanboard/kanboard#2324]
I am working on developing a SSO plugin for authentication.
The way SAML works is that there are IdentityProviders (IDP) and ServiceProviders (SP), that exchange a handshake and the IDP provides user details to the SP.

In this case Kanboard would be a SP. I've gotten to the point that I made a controller that gets the details in /ssologin $POST['email'], $POST['username'], however I need some help in order to finish it up.

How can I use the PreAuthenticationProviderInterface in order to:
1. check if user with $POST['email'] exists already and consequently authenticate them
2. create a user based on the received $POST['email'], $POST['username'] and consequently authenticate them

I've gone over the documentation but unable to put it together just yet...
Basically it's creating a user from the $POST['email'] response and in case they exist, authenticating them where I am stuck.

Thanks for the great software! Hoping to contribute with this plugin soon.

**Comment by fguillot** [comment]
What is the authentication flow for a user? I assume that is similar to OAuth2, right?

**Comment by trajche** [comment]
Yes, you are right, I guess it's similar to OAuth2. Basically, redirect to a third party link, return back to service with user details contained in $POST[] request. I will have a look at the OAuth pluings you have made.

In which cases would PreAuthentication interface be used then? Or is that just to make sure that there is a session for each request?

Thanks!

**Comment by fguillot** [comment]
You could use the user metadata to store the unique user id so you don't have to alter the schema. Or choose the hard way, create a sql migration in your plugin to add a new column.

I was also planning to write a SAML authentication plugin for Kanboard but I didn't start yet due to lack of time. Let me know if you need some modification in the core.

**Comment by trajche** [comment]
Hmm - I've been checking out the OAuth plugins in order to try and use the same class, however I don't think it would work as it's not exactly the same. SAML offers different outputs in a $POST[] request depending on how the authentication has been setup from the IdentityProvider side...

To put it down to the simplest question...

If I had a $POST['name'] and $POST['email'] response ONLY... how would I go about registering (if account doesn't exist) and consequently logging in that user?

**Comment by trajche** [comment]
I have released this as a SAML plugin for Kanboard: https://github.com/trajche/SamlAuth
Will do a pull request on plugins page when I have time.
